packages/core/src/components/overlay2/overlay2.tsx

```diff
@@ -0,0 +1,710 @@
+
+    /* An empty, keyboard-focusable div at the end of the Overlay content */
+    const endFocusTrapElement = React.useRef<HTMLDivElement>(null);
+
+    /**
+     * Locally-generated DOM ref for a singleton child element.
+     * This is only used iff the user does not specify the `childRef` or `childRefs` props.
+     */
+    const localChildRef = React.useRef<HTMLElement>(null);
+
+    const bringFocusInsideOverlay = React.useCallback(() => {
+        // always delay focus manipulation to just before repaint to prevent scroll jumping
+        return requestAnimationFrame(() => {
+            // container element may be undefined between component mounting and Portal rendering
+            // activeElement may be undefined in some rare cases in IE
+            const container = getRef(containerElement);
+            const activeElement = getActiveElement(container);
+
+            if (container == null || activeElement == null || !isOpen) {
+                return;
+            }
+
+            // Overlay2 is guaranteed to be mounted here
+            const isFocusOutsideModal = !container.contains(activeElement);
+            if (isFocusOutsideModal) {
+                getRef(startFocusTrapElement)?.focus({ preventScroll: true });
+                setIsAutoFocusing(false);
+            }
+        });
+    }, [isOpen]);
+
+    /**
+     * When multiple Overlays are open, this event handler is only active for the most recently
+     * opened one to avoid Overlays competing with each other for focus.
+     */
+    const handleDocumentFocus = React.useCallback(
+        (e: FocusEvent) => {
+            // get the actual target even in the Shadow DOM
+            // see https://github.com/palantir/blueprint/issues/4220
+            const eventTarget = e.composed ? e.composedPath()[0] : e.target;
+            const container = getRef(containerElement);
+            if (
+                enforceFocus &&
+                container != null &&
+                eventTarget instanceof Node &&
+                !container.contains(eventTarget as HTMLElement)
+            ) {
+                // prevent default focus behavior (sometimes auto-scrolls the page)
+                e.preventDefault();
+                e.stopImmediatePropagation();
+                bringFocusInsideOverlay();
+            }
+        },
+        [bringFocusInsideOverlay, enforceFocus],
+    );
+
+    const id = useOverlay2ID();
+    const instance = React.useMemo<OverlayInstance>(
+        () => ({
+            bringFocusInsideOverlay,
+            containerElement,
+            handleDocumentFocus,
+            id,
+            props: {
+                autoFocus,
+                enforceFocus,
+                hasBackdrop,
+                usePortal,
+            },
+        }),
+        [autoFocus, bringFocusInsideOverlay, enforceFocus, handleDocumentFocus, hasBackdrop, id, usePortal],
+    );
+
+    React.useEffect(() => {
+        setRef(ref, instance);
```

Comment:
consider [`useImperativeHandle`](https://react.dev/reference/react/useImperativeHandle) to do this instead